File: ui-selenium/src/main/java/org/aeautomation/utils/ConfigReader.java
package org.aeautomation.utils;

import java.io.IOException;
import java.io.InputStream;
import java.util.Objects;
import java.util.Properties;
import java.util.stream.Stream;

/**
 * Utility class responsible for loading framework configurations.
 * Implements a fallback mechanism: CLI system properties (-D key=value)
 * take precedence over properties defined in config.properties.
 */
public final class ConfigReader {

    private static final Properties properties = new Properties();
    private static final String CONFIG_FILE_PATH = "config.properties";

    // Static initializer to load properties once when the class is loaded into memory
    static {
        try (InputStream input = ConfigReader.class.getClassLoader()
                .getResourceAsStream(CONFIG_FILE_PATH)) {

            if (input == null) {
                throw new IllegalStateException("Configuration file '" + CONFIG_FILE_PATH + "' not found in classpath.");
            }
            properties.load(input);

        } catch (IOException e) {
            throw new RuntimeException("Failed to load configuration file: " + CONFIG_FILE_PATH, e);
        }
    }

    private ConfigReader() {}

    /**
     * Property getter that resolves configuration values based on strict priority.
     *
     * @param key            The property key to resolve
     * @param optionalValues Optional runtime parameter fallbacks
     * @return Resolved property value string
     * @throws IllegalArgumentException if the property key is missing across all configuration sources
     */
    public static String getProperty(String key, String... optionalValues) {
        return Stream.concat(
                        Stream.of(System.getProperty(key), properties.getProperty(key)),
                        Stream.of(optionalValues)
                )
                .filter(Objects::nonNull)
                .map(String::trim)
                .filter(val -> !val.isEmpty())
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException(
                        "Property key '" + key + "' was not found in System properties, config.properties, or optional parameters!"));
    }

    /**
     * Gets the global explicit wait timeout in seconds.
     *
     * @return Timeout duration in seconds (defaults to 10 seconds if unspecified)
     */
    public static int getTimeout() {
        try {
            return Integer.parseInt(getProperty("timeout"));
        } catch (NumberFormatException e) {
            System.err.println("Invalid 'timeout' value in configuration. Defaulting to 10 seconds.");
            return 10;
        }
    }

    /**
     * Retrieves the target browser for test execution, allowing an optional TestNG XML override.
     *
     * @param xmlBrowser Browser parameter optionally injected from testng.xml
     * @return Resolved browser string
     */
    public static String getBrowser(String xmlBrowser) {
        return getProperty("browser", xmlBrowser);
    }

    /**
     * Retrieves the target browser for test execution using the standard priority hierarchy.
     *
     * @return Resolved browser string
     */
    public static String getBrowser() {
        return getProperty("browser", "CHROME");
    }

    /**
     * Retrieves the target application base URL for navigating test environments.
     *
     * @return Base URL string
     */
    public static String getBaseUrl() {
        return getProperty("baseUrl");
    }

    /**
     * Determines whether browser execution should run in headless mode.
     *
     * @return true if 'headless' property is set to "true", otherwise false
     */
    public static boolean isHeadless() {
        return Boolean.parseBoolean(getProperty("headless", "false"));
    }

    /**
     * Convenience method to retrieve a property parsed as a boolean.
     */
    public static boolean getBoolean(String key, String... optionalValues) {
        return Boolean.parseBoolean(getProperty(key, optionalValues));
    }

    /**
     * Convenience method to retrieve a property parsed as an integer (e.g. timeouts).
     */
    public static int getInt(String key, String... optionalValues) {
        return Integer.parseInt(getProperty(key, optionalValues));
    }
}

File: ui-playwright/utils/ConfigReader.ts
import * as dotenv from 'dotenv';
import Logger from './Logger';

dotenv.config();

/**
 * Utility class responsible for loading framework configurations.
 * Implements a fallback mechanism: environment variables (process.env) take precedence
 * over values defined in the .env file.
 *
 * All configuration keys must be defined in .env or passed as environment variables.
 * See .env.example for the full list of required keys.
 */
class ConfigReader {

    /**
     * Resolves a configuration value by key.
     * Priority: process.env (CLI) → .env file → optional defaultValue
     *
     * @param key - Environment variable key
     * @param defaultValue - Optional fallback value
     * @returns Resolved string value
     * @throws Error if key is missing and no default is provided
     */
    static get(key: string, defaultValue?: string): string {
        const value = process.env[key]?.trim();

        if (value !== undefined && value !== '') {
            Logger.debug(`Resolved config [${key}] -> '${value}'`);
            return value;
        }

        if (defaultValue !== undefined) {
            Logger.debug(`Config [${key}] not set, using fallback -> '${defaultValue}'`);
            return defaultValue;
        }

        Logger.error(`Required config key '${key}' was not found in process.env or .env file`);
        throw new Error(`Required config key '${key}' is missing from environment and .env file`);
    }

   /**
     * Retrieves the target application base URL (`BASE_URL`).
     * @returns Base URL string
     */
    static getBaseUrl(): string {
        return this.get('BASE_URL');
    }

    /**
     * Checks if browser execution should run in headless mode (`HEADLESS`).
     * @returns True if headless mode is enabled, false if set to 'false'
     */
    static isHeadless(): boolean {
        return this.get('HEADLESS').toLowerCase() !== 'false';
    }

    /**
     * Retrieves the global test timeout in milliseconds (`TIMEOUT`).
     * @returns Timeout duration in milliseconds
     * @throws Error if TIMEOUT configuration value is not a valid numeric string
     */
    static getTimeout(): number {
        const timeoutStr = this.get('TIMEOUT');
        const parsed = parseInt(timeoutStr, 10);
        
        if (isNaN(parsed)) {
            Logger.error(`Invalid numeric TIMEOUT configuration: '${timeoutStr}'`);
            throw new Error(`Config key 'TIMEOUT' must be a valid integer, received: '${timeoutStr}'`);
        }
        return parsed;
    }

    /**
     * Retrieves the domain name used for dynamic test email generation (`TEST_EMAIL_DOMAIN`).
     * @returns Email domain string (e.g., 'example.com')
     */
    static getEmailDomain(): string {
        return this.get('TEST_EMAIL_DOMAIN');
    }

    /**
     * Retrieves the configured framework logging threshold (`LOG_LEVEL`).
     * @returns Log level string ('debug' | 'info' | 'warn' | 'error')
     */
    static getLogLevel(): string {
        return this.get('LOG_LEVEL');
    }

    /**
     * Retrieves the explicit wait timeout for element visibility and assertions in milliseconds (`EXPLICIT_WAIT_TIMEOUT`).
     * @returns Explicit wait duration in milliseconds (defaults to 10000)
     */
    static getExplicitWaitTimeout(): number {
        return parseInt(this.get('EXPLICIT_WAIT_TIMEOUT', '10000'), 10);
    }
}

export default ConfigReader;